### PR TITLE
feat: add `strategy="quantile"` in KBinsDiscretizer

### PR DESCRIPTION
**Pull request by ashleyxuu**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes internal #310685445 🦕

### CODE REVIEW

File: bigframes/ml/preprocessing.py [hunk 1]
```diff
@@ -337,15 +333,31 @@ def _compile_to_sql(
...
+                    self._base_sql_generator.ml_bucketize(
+                        column, array_split_points[column], f"kbinsdiscretizer_{column}"
+                    ),
+                    f"kbinsdiscretizer_{column}",
+                )
+                for column in columns
+            ]
+
+        if self.strategy == "quantile":
+            # quantile_size = 1 / self.n_bins
```
Comment (GarrettWu) [inline bigframes/ml/preprocessing.py:347]: remove?

File: bigframes/ml/preprocessing.py [hunk 2]
```diff
@@ -356,11 +368,16 @@ def _parse_from_sql(cls, sql: str) -> tuple[KBinsDiscretizer, str]:
...
         Returns:
             tuple(KBinsDiscretizer, column_label)"""
-        s = sql[sql.find("(") + 1 : sql.find(")")]
-        array_split_points = s[s.find("[") + 1 : s.find("]")]
-        col_label = s[: s.find(",")]
-        n_bins = array_split_points.count(",") + 2
-        return cls(n_bins, "uniform"), col_label
+        if sql.startswith("ML.QUANTILE_BUCKETIZE"):
+            s = sql[sql.find("(") + 1 : sql.find(")")]
+            col_label, num_bins = s.split(", ")
```
Comment (GarrettWu) [inline bigframes/ml/preprocessing.py:373]: split by ", " is kinda dangerous. What if the SQL doesn't contain a space. 

What you actually want is split(",") with .strip on each str.

Comment (GarrettWu) [reply bigframes/ml/preprocessing.py:373]: ```s = sql[sql.find("(") + 1 : sql.find(")")]```
Can be shared.

Comment (ashleyxuu) [reply bigframes/ml/preprocessing.py]: Done.

File: bigframes/ml/preprocessing.py (hunk 1 above)
Comment (ashleyxuu) [reply bigframes/ml/preprocessing.py]: Done.

File: bigframes/ml/preprocessing.py [hunk 3]
```diff
@@ -368,14 +364,14 @@ def _parse_from_sql(cls, sql: str) -> tuple[KBinsDiscretizer, str]:
```
Comment (GarrettWu) [inline bigframes/ml/preprocessing.py:364]: update the comment.

File: bigframes/ml/preprocessing.py [hunk 4]
```diff
@@ -337,15 +333,27 @@ def _compile_to_sql(
...
+                (
+                    self._base_sql_generator.ml_bucketize(
+                        column, array_split_points[column], f"kbinsdiscretizer_{column}"
+                    ),
+                    f"kbinsdiscretizer_{column}",
+                )
+                for column in columns
+            ]
+
+        if self.strategy == "quantile":
```
Comment (GarrettWu) [inline bigframes/ml/preprocessing.py:346]: Can use
```
elif:

else:
  raise ValueError
```

So if user gives the wrong message, then they will get exceptions.

**Review by GarrettWu — APPROVED** [review]

**Pull request merged**